Added error messages for inconsistent systems in solve_for_reaction_loads

saiudayagiri opened the pull request:
#### Added error messages for inconsistent systems in solve_for_reaction_loads

#### References to other Issues 
Fixes #28346 

#### what is fixed or changed
An logic is written in solve_for_reaction_loads to find inconsistent systems before passing into linsolve as linsolve doesnt return any solution leading to index error 
**Changes Made:**

- Conditional consistency if an condition makes the system solvable it raises the condition saying user "this system is solvalble under this condition only" 
- for inconsistent systems it raises inconsistent system detected


**Test Cases Added:**
- Inconsistent systems
- conditional consistency

#### Other comments
No functional impact outside of clearer error reporting.

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics

  * Improved error reporting to avoid confusion for users for inconsistent systems which before  simply returned key index error not giving any msg why the system fails.
<!-- END RELEASE NOTES -->

oscarbenjamin commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -974,8 +974,52 @@ def solve_for_reaction_loads(self, *reactions):
+
+        unknowns  = (C3, C4) + reactions + rotation_jumps + deflection_jumps
+
+        try:
+            sol_tuple = linsolve(equations, unknowns).args[0]
+
+        except Exception as err:
+            support_syms = {load[0] for load in self._support_as_loads}
```
Catching Exception will catch too many unexpected exception types. When using `try/except` you should be specific about what kinds of exception to catch.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -939,6 +939,12 @@ def solve_for_reaction_loads(self, *reactions):
+        if not reactions:
+            raise ValueError("No symbols supplied to solve_for_reaction_loads().")
+
+        if len(set(reactions)) != len(reactions):
```
This doesn't trigger for only duplicate symbols

saiudayagiri replied:
@Tom-van-Woudenberg have gone through a bit of revisit and I don't see any other failures.

Tom-van-Woudenberg replied:
Ah, you're right. Ignore my comment

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -974,8 +980,24 @@ def solve_for_reaction_loads(self, *reactions):
+        rank_Ab = A.row_join(b).rank()
+
+        if rank_A != rank_Ab:
+            raise ValueError("System is statically inconsistent. No Solution.")
```
I think statically inconsistent can be clarified a bit, it's not a term everyone knows.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -974,8 +980,24 @@ def solve_for_reaction_loads(self, *reactions):
+        try:
+            solutions = linsolve(equations, unknowns)
+            if not solutions:
+                raise ValueError("Could not find a solution.")
```
I don't see a test for this

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -974,8 +980,24 @@ def solve_for_reaction_loads(self, *reactions):
+            solutions = linsolve(equations, unknowns)
+            if not solutions:
+                raise ValueError("Could not find a solution.")
+            solution = list(solutions.args[0])
```
What will this be when it cannot find a solution?

Tom-van-Woudenberg replied:
Is this related to mechanisms and indeterminate case? Maybe add a comment?

Tom-van-Woudenberg reviewed (changes requested):
Nice start, I've added some comments

saiudayagiri commented:
> Nice start, I've added some comments

Thanks for the feedback, @Tom-van-Woudenberg ! I’ve reworked the implementation to address your comments: Could you take another look and let me know if there’s anything else to refine?

oscarbenjamin commented:
> real-world mechanisms that aren’t fully constrained are inherently unstable and shouldn’t admit undetermined reaction parameters.

Being statically indeterminate does not imply instability. It just means that the basic rigid body external force and moment balance equations do not uniquely determine the reaction forces. Usually the rigid body assumption needs to be relaxed to determine the reactions.

A four-legged chair resting on the floor is statically indeterminate but not unstable. In 2D the same is true of a three-legged chair (e.g. a horizontal beam resting on three vertical supports).

saiudayagiri commented:
> > real-world mechanisms that aren’t fully constrained are inherently unstable and shouldn’t admit undetermined reaction parameters.
> 
> Being statically indeterminate does not imply instability. It just means that the basic rigid body external force and moment balance equations do not uniquely determine the reaction forces. Usually the rigid body assumption needs to be relaxed to determine the reactions.
> 
> A four-legged chair resting on the floor is statically indeterminate but not unstable. In 2D the same is true of a three-legged chair (e.g. a horizontal beam resting on three vertical supports).

@oscarbenjamin The solve_for_reaction_loads still solves both determinate and indeterminate systems as before.

saiudayagiri commented:
> > In the previous implementation, linsolve would often return parametric solutions for under-constrained (mechanism) systems—introducing free symbols to represent the infinite family of equilibrium states. In the current version, that behavior has been modified: such parametric solutions are no longer produced
> 
> Can you explain an example of this?

@oscarbenjamin  Now there is no behavior change — the solver handles inconsistent systems by returning an error that shows users which equations contradict each other and provides no solution. The cases I consider as mechanisms that are failing before  are beams with a single pin or roller support, which previously caused key‐index errors. Now, all of those fall under a single case “inconsistent,” as they return an empty solution set due to contradictory equations.

saiudayagiri commented:
@Tom-van-Woudenberg @Upabjojr @oscarbenjamin I’ve just pushed an update to solve_for_reaction_loads that replaces the previous rank check with an RREF-based consistency test. Could you take a look when you have a moment?

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_beam.py:
```diff
@@ -307,6 +307,105 @@ def test_statically_indeterminate():
+    b.apply_load(R1, 0, -1)
+    b.apply_load(R1, 10, -1)
+    b.bc_deflection = [(0, 0), (10, 0)]
+    with raises(ValueError, match="Duplicate symbols supplied to solve_for_reaction_loads()."):
```
This doesn't have to be an error. Someone might want to enforce reactions forces being equal by assigning the same symbol for them.

Tom-van-Woudenberg replied:
Ah, okay. But then it makes sense so do the condition check when defining the load, instead of when solving

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/beam.py:
```diff
@@ -974,8 +980,35 @@ def solve_for_reaction_loads(self, *reactions):
+            if row_sum == 0 and rhs != 0:
+                if not rhs.is_number:
+                    raise ValueError(
+                            f"The system is only solvable with symbolic constraint"
+                            f" {rhs} = 0.substitute to proceed"
```
I don't fully understand what is meant with this error message. Seems related to https://github.com/sympy/sympy/pull/28113#issuecomment-3015473181 response, in which it is mentioned that for a zero rhs the lhs can have only zeros. However, how would a user 'substitute rhs = 0' for the beam? I'd prefer an error message in terms of beam 'behaviour'.

Tom-van-Woudenberg replied:
@oscarbenjamin, might provide his opinion on this too

Tom-van-Woudenberg replied:
Hmm, I'm curious to @oscarbenjamin's opinion, but suggesting to change the load might trick an user into doing that wihtout realizing it changes their system. Furthermore, the suggestion is not always a solution, see https://github.com/sympy/sympy/pull/28113#discussion_r2191700693

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_beam.py:
```diff
@@ -307,6 +307,105 @@ def test_statically_indeterminate():
+    with raises(ValueError, match="This means your supports/BCs generate contradictory or insufficient constraints."):
+        b.solve_for_reaction_loads(R1)
+
+    # conditional consistent system of equations with sliding hinge
+    E, I = Symbol('E'), Symbol('I')
+    b = Beam(8, E, I)
+    R1=b.apply_support(0,'roller')
+    b.apply_sliding_hinge(4)
+    b.apply_load(5, 8, -1)
+    R1 = Symbol('R1')
+    with raises(ValueError, match="The system is only solvable with symbolic constraint"):
+        b.solve_for_reaction_loads(R1)
```
Related to my comment on line 996-997 of beam.py.

The reduced echelon form of this system is:

```
Matrix([[0, 1, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
Matrix([[0], [R_0 + 5], [8*R_0], [R_0]])
```

While, the valueerror is:

```
The system is only solvable with symbolic constraint R_0 + 5 = 0.substitute to proceed
```

This additional constraint would not match row 3 and 4. so is rather confusing.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_beam.py:
```diff
@@ -307,6 +307,105 @@ def test_statically_indeterminate():
+    with raises(ValueError, match="The system is only solvable with symbolic constraint"):
+        b.solve_for_reaction_loads(R1)
+
+    b = Beam(10, E, I)
+    b.apply_support(0, 'fixed')
+    b.apply_rotation_hinge(5)
+    b.apply_support(10, 'fixed')
+    b.apply_load(-P, 5, -1)
+    with raises(ValueError, match="The system is only solvable with symbolic constraint"):
+        b.solve_for_reaction_loads(R1)
```
This system should be solveable. It is a statically indeterminant beam

saiudayagiri replied:
yes here this is solvable yet failing because i missed a check for minimum symbols to be provided .here eventhough the system expects four symbols from two fixed supports we only added one reaction symbol. an check is added for this now

Tom-van-Woudenberg replied:
I'd except:
```
{R_0: P/2 + ... M_0 + ... R_10 + ...M_10}
```

with some factor in place of every ... . I remember a similar discussion with @lfverlaan, see 2.2.2 of https://repository.tudelft.nl/file/File_7a8feba9-ad25-48f3-9263-88c3749abca8?preview=1

Tom-van-Woudenberg replied:
The second approach is not wrong, so I wouldn't recommend changing it.

> so users must explicitly supply them. This can lead to confusion or underdetermined systems if not done carefully.

Users might want a result for reaction symbols in terms of each other, independent of whether the system is underdetermined or not.

Tom-van-Woudenberg replied:
```python
E, I = symbols('E, I')
P = Symbol('P', positive=True)
b = Beam(10, E, I)
b.apply_support(0, 'fixed')
b.apply_rotation_hinge(5)
b.apply_support(10, 'fixed')
b.apply_load(-P, 5, -1)
R_0 = Symbol('R_0')
b.solve_for_reaction_loads(R_0)
```
gives
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line=9)
      7 b.apply_load(-P, 5, -1)
      8 R_0 = Symbol('R_0')
----> 9 b.solve_for_reaction_loads(R_0)

File c:\Users\tomvanwoudenbe\git\sympy\sympy\physics\continuum_mechanics\beam.py:950, in Beam.solve_for_reaction_loads(self, *reactions)
    948 required = len(self.bc_deflection) + len(self.bc_slope)
    949 if len(reactions) < required:
--> 950    raise ValueError(f"Expected at least {required} reaction symbols (based on deflection and slope BCs), but got {len(reactions)}.")
    953 x = self.variable
    954 l = self.length

ValueError: Expected at least 4 reaction symbols (based on deflection and slope BCs), but got 1.
```
Which is the error message you added in https://github.com/sympy/sympy/commit/f776f7ee750675b7777a6718b49826c70f087211 while I'd expect an equation for R_0 in terms of P, M_0, R_10 and M_10

Commit:
inconsistent systems
